lib/site-inspector.rb

```diff
@@ -127,37 +134,213 @@ def to_json
-      :google_apps => google_apps?,
-      :cloud_provider => cloud_provider,
-      :server => server,
-      :cms => cms,
-      :analytics => analytics,
-      :javascript => javascript,
-      :advertising => advertising,
-      :slash_data => slash_data?,
-      :slash_developer => slash_developer?,
-      :data_dot_json => data_dot_json?,
-      :click_jacking_protection => click_jacking_protection?,
-      :content_security_policy => content_security_policy?,
-      :xss_protection => xss_protection?,
-      :secure_cookies => secure_cookies?,
-      :strict_transport_security => strict_transport_security?,
-      :headers => headers
+      https: {
+        www: https_www,
+        root: https_root
+      },
+      http: {
+        www: http_www,
+        root: http_root
+      }
     }
   end
+
+  # State of affairs at a particular endpoint.
+  def http_endpoint(ssl, www)
+    details = {}
+
+    # Don't follow redirects for first ping.
+    response = request(ssl, www, false)
+
+
+    # For HTTPS: examine the full range of possibilities.
+    if ssl
+      if response.return_code == :ok
+        details[:https_valid] = true
+
+      # Bad certificate chain.
+      elsif response.return_code == :ssl_cacert
+        details[:https_valid] = false
+        details[:https_bad_chain] = true
+        response = request(ssl, www, false, false, true)
+        # Bad everything.
+        if response.return_code == :peer_failed_verification
+          details[:https_bad_name] = true
+          response = request(ssl, www, false, false, false)
+        end
+      # Bad hostname.
+      elsif response.return_code == :peer_failed_verification
+        details[:https_valid] = false
+        details[:https_bad_name] = true
+        response = request(ssl, www, false, true, false)
+        # Bad everything.
+        if response.return_code == :ssl_cacert
+          details[:https_bad_chain] = true
+          response = request(ssl, www, false, false, false)
+        end
+      end
+    end
+
+    # If we ended up with a failure, return it.
+    details[:status] = response.response_code
+    return details if response.response_code == 0
+
+    headers = Hash[response.headers.map{ |k,v| [k.downcase,v] }]
+    # details[:headers] = headers
+
+
+    # HSTS only takes effect when delivered over valid HTTPS.
+    details[:hsts] = !!(ssl and details[:https_valid] and headers["strict-transport-security"])
+    details[:hsts_header] = headers["strict-transport-security"]
+
+
+    # If it's a redirect, go find the ultimate response starting from this combo.
+    redirect_code = response.response_code.to_s.start_with?("3")
+    location_header = headers["location"]
+    if redirect_code and location_header
+      details[:redirect] = true
+
+      ultimate_response = request(ssl, www, true, !details[:https_bad_chain], !details[:https_bad_name])
+      uri_original = URI(ultimate_response.request.url)
+      uri_eventual = URI(ultimate_response.effective_url)
```

Comment:
For the `effective_url` field to be filled in during the test suite, VCR must be hooked into `typheous`, not `webmock`, so that redirects information is successfully captured. This change (and the regenerated cassettes it required) are present in #23 already, and was generally helpful to speed up the test suite and kill unnecessary network requests.